File: homework4.py
'''
	Austin Fouch
	CMPS 367 Python Programming
	Homework #4
	homework3.py
		This program takes in two files as command line arguments and
		calculates the Levenshtein Distance of the strings contained
		in the files. Also displays the aligned strings. 
'''
import sys	# for command line args
import time	# for program exec time

# load_file()
#	returns the first string in a file
def load_file(filename):
	content = []
	f = open(filename, 'r')
	content = [x.strip('\n') for x in f.readlines()]
	f.close()
	return content[0]
# load_file()

# get_input()
#	returns tuple that contains the strings from the input files
def get_input():
	str1 = load_file(sys.argv[1])	# file 1 as str
	str2 = load_file(sys.argv[2])	# file 2 as str
	print("Input Sequences")
	print("-" * 30)
	print(str1)
	print(str2)
	print("-" * 30)
	return (str1, str2)
# get_input()

# calc_Levenshtein_Distance()
#	creates matrix based on the input strings, calculates the Levenshtein
#	Distance between the strings, and prints the strings out, aligned
def calc_Levenshtein_Distance():
	start = time.time()
	tmp = get_input()
	S = tmp[0]			# first file as string
	T = tmp[1]			# second file as string
	M = len(S) + 1
	N = len(T) + 1
	D = [[0 for element in range(M)] for element in range(N)]
	E = [[0 for element in range(M)] for element in range(N)]
	''' inserting into D and E '''
	for i in range(0, M):
		D[0][i] = i
		for j in range(0, N):
			D[j][0] = j
	for i in range(1,N):
		for j in range(1,M):		
			if S[j-1] == T[i-1]:
				D[i][j] = D[i-1][j-1]
				E[i][j] = 0
			else:
				x = D[i-1][j] + 1
				y = D[i][j-1] + 1
				z = D[i-1][j-1] + 1
				D[i][j] = min((x,y,z))
				if x == min((x,y,z)):
					E[i][j] = 1
				elif y == min((x,y,z)):
					E[i][j] = 2
				elif z == min((x,y,z)):
					E[i][j] = 0
	''' creating result strings '''
	S_RESULT = ""
	T_RESULT = ""
	i = N - 1
	j = M - 1
	while i > 0 or j > 0:
		if E[i][j] == 0:
			S_RESULT = S[j-1] + S_RESULT
			T_RESULT = T[i-1] + T_RESULT
			i -= 1
			j -= 1
		if E[i][j] == 1:
			S_RESULT = '-' + S_RESULT
			T_RESULT = T[i-1] + T_RESULT
			i -= 1
		if E[i][j] == 2:
			S_RESULT = S[j-1] + S_RESULT
			T_RESULT = '-' + T_RESULT
			j -= 1
	end = time.time()
	''' print result strings, edit distance, and execution time '''
	print("Aligned Sequences")
	print("-"*30)
	print(S_RESULT)
	print(T_RESULT)
	print("-"*30)
	print("The minimum edit distance is", D[-1][-1])
	print("Completed in", float(end-start), "seconds")
# calc_Levenshtein_Distance()

# Main
calc_Levenshtein_Distance()
# Main

File: homework4.cpp
#include <ctime>
#include <iostream>
#include <fstream>
#include <streambuf>
#include <string>
#include <vector>
#include <algorithm>
#include <iomanip>
using namespace std;

void calc_Levenshtein_Distance(const string &a, const string &b)
{
    /* copy strings from args, get lengths, create the arrays */
    string S = a;
    string T = b;
    int M = S.length() + 1;
    int N = T.length() + 1;
    int D[N][M];
    int E[N][M];
    int i,j;
    /* fill arrays */
    for(i = 0; i < N; i++)
        for(j = 0; j < M; j++)
        {
            D[i][j] = 0;
            E[i][j] = 0;
        }
    for(i = 0; i < M; i++)
        D[0][i] = i;
    for(j = 0; j < N; j++)
        D[j][0] = j;
    /* get the edit distance */
    int x, y, z;
    for(i = 1; i < N; i++)
        for(j = 1; j < M; j++)
            if(S[j-1] == T[i-1])
            {
                D[i][j] = D[i-1][j-1];
            } 
            else
            {
                x = D[i-1][j] + 1;
                y = D[i][j-1] + 1;
                z = D[i-1][j-1] + 1;
                D[i][j] = min(x, min(y, z));
                if(D[i][j] == x)
                    E[i][j] = 1;
                else if(D[i][j] == y)
                    E[i][j] = 2;
                else if(D[i][j] == z)
                    E[i][j] = 0;
            }
    /* creating result strings */
    string S_RESULT = "";
    string T_RESULT = "";
    i = N - 1;
    j = M - 1;
    while(i > 0 || j > 0)
    {
        if(E[i][j] == 0)
        {
            S_RESULT.insert(0, 1, S[j-1]);
            T_RESULT.insert(0, 1, T[i-1]);
            i -= 1;
            j -= 1;
        }
        if(E[i][j] == 1)
        {
            S_RESULT.insert(0, 1, '-');
            T_RESULT.insert(0, 1, T[i-1]);
            i -= 1;
        }
        if(E[i][j] == 2)
        {
            S_RESULT.insert(0, 1, S[j-1]);
            T_RESULT.insert(0, 1, '-');
            j -= 1;
        }
    }
    /* printing results */
    cout << "Aligned Sequences" << endl;
    cout << string(30, '-') << endl;
    cout << S_RESULT << endl;
    cout << T_RESULT << endl;
    cout << string(30, '-') << endl;
    cout << "The minimum edit distance is " << D[N-1][M-1] << endl;
}

int main(int argv, char ** argc) {
    /* start timer */
    clock_t start, end;
    start = clock();
    /* parse files into strings */
    ifstream f1(argc[1]);
    string str1((istreambuf_iterator<char>(f1)),
    			 istreambuf_iterator<char>());
    ifstream f2(argc[2]);
    string str2((istreambuf_iterator<char>(f2)),
    			 istreambuf_iterator<char>());
    /* print and pass results to edit distance calculator */
    cout << "Input Sequences" << endl;
    cout << string(30, '-') << endl;
    cout << str1 << endl;
    cout << str2 << endl;
    cout << string(30, '-') << endl;
    calc_Levenshtein_Distance(str1, str2);
    cout << string(30, '-') << endl;
    /* print time */
    end = clock();
    float diff = ((float)end - (float)start);
    float seconds = diff / CLOCKS_PER_SEC;
    cout << "Completed in " << setprecision(10) << seconds << " seconds." << endl;
    return 0;
}